File: go/problem_48.go
package main

type TreeNode48 struct {
	val         string
	left, right *TreeNode48
}

func buildTree(preorder, inorder []string) *TreeNode48 {
	if len(preorder) == 0 {
		return nil
	}
	root := &TreeNode48{val: preorder[0]}
	idx := 0
	for i, v := range inorder {
		if v == preorder[0] {
			idx = i
			break
		}
	}
	root.left = buildTree(preorder[1:1+idx], inorder[:idx])
	root.right = buildTree(preorder[1+idx:], inorder[idx+1:])
	return root
}

func inorderTraversal48(root *TreeNode48) []string {
	if root == nil {
		return nil
	}
	result := inorderTraversal48(root.left)
	result = append(result, root.val)
	result = append(result, inorderTraversal48(root.right)...)
	return result
}

func problem48() {
	println("Problem 48:")
	pre := []string{"a", "b", "d", "e", "c", "f", "g"}
	in := []string{"d", "b", "e", "a", "f", "c", "g"}
	root := buildTree(pre, in)
	got := inorderTraversal48(root)
	assertSliceEqual(got, in)
	println("All Tests Passed")
}

File: rust/problem_48.rs
#[derive(Debug)]
pub struct TreeNode48 {
    pub val: String,
    pub left: Option<Box<TreeNode48>>,
    pub right: Option<Box<TreeNode48>>,
}

pub fn build_tree(preorder: &[&str], inorder: &[&str]) -> Option<Box<TreeNode48>> {
    if preorder.is_empty() { return None; }
    let root_val = preorder[0];
    let idx = inorder.iter().position(|&v| v == root_val).unwrap();
    Some(Box::new(TreeNode48 {
        val: root_val.to_string(),
        left:  build_tree(&preorder[1..1+idx], &inorder[..idx]),
        right: build_tree(&preorder[1+idx..], &inorder[idx+1..]),
    }))
}

fn inorder48(root: &Option<Box<TreeNode48>>, out: &mut Vec<String>) {
    if let Some(n) = root { inorder48(&n.left, out); out.push(n.val.clone()); inorder48(&n.right, out); }
}

pub fn problem48() {
    println!("Problem 48:");
    let pre = vec!["a","b","d","e","c","f","g"];
    let ino = vec!["d","b","e","a","f","c","g"];
    let root = build_tree(&pre, &ino);
    let mut out = vec![];
    inorder48(&root, &mut out);
    let exp: Vec<String> = ino.iter().map(|s| s.to_string()).collect();
    assert_eq!(out, exp);
    println!("All Tests Passed");
}
